src/usr.bin/encrypt/encrypt.1
.\"	$MirOS$
.\"	$OpenBSD: encrypt.1,v 1.18 2003/06/14 23:19:18 millert Exp $
.\"
.\" Copyright (c) 1996, Jason Downs.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR(S) ``AS IS'' AND ANY EXPRESS
.\" OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
.\" DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR(S) BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
.\" CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd $Mdocdate$
.Dt ENCRYPT 1
.Os
.Sh NAME
.Nm encrypt
.Nd encrypt passwords from the command line or standard input
.Sh SYNOPSIS
.Nm encrypt
.Fl b Ar rounds
.Op Fl p | Ar string
.Nm encrypt
.Op Fl c Ar class
.Op Fl p | Ar string
.Nm encrypt
.Fl m
.Op Fl p | Ar string
.Nm encrypt
.Fl S Ar salt
.Op Fl p | Ar string
.Nm encrypt
.Fl s Ar salt
.Op Fl p | Ar string
.Nm encrypt
.Fl k
.Nm makekey
.Sh DESCRIPTION
.Nm
prints the encrypted form of
.Ar string
to the standard output.
This is mostly useful for encrypting passwords from within scripts.
.Pp
The options are as follows:
.Bl -tag -width xbxrounds
.It Fl k
Run in
.Nm makekey
compatible mode; a single combined key and salt are read from standard
input and the DES encrypted result is written to standard output without a
terminating newline.
.It Fl b Ar rounds
Encrypt the string using Blowfish hashing with the specified
.Ar rounds .
.It Fl c Ar class
Use the cipher type specified in the given user login class.
See
.Xr login.conf 5
for more information.
.It Fl m
Encrypt the string using MD5.
.It Fl p
Prompt for a single string with echo turned off.
.It Fl S Ar salt
Encrypt the string with the specified
.Ar salt ,
using the algorithm specified by
.Ar salt .
.It Fl s Ar salt
Encrypt the string using DES, with the specified
.Ar salt .
.El
.Pp
If no
.Ar string
is specified,
.Nm
reads one string per line from standard input, encrypting each one
with the chosen algorithm from above.
In the case where no specific algorithm or specific user login class was given
as a command line option, the algorithm specified in the default class in
.Pa /etc/login.conf
will be used.
.Pp
For MD5 and Blowfish as well as any algorithm determined from the login
.Ar class ,
a new random salt is automatically generated for each password, except if the
.Fl S
option is used.
.Pp
Specifying the
.Ar string
on the command line should be discouraged; using the
standard input is more secure.
.Sh FILES
.Bl -tag -width /etc/login.conf -compact
.It Pa /etc/login.conf
.El
.Sh SEE ALSO
.Xr crypt 3 ,
.Xr login.conf 5
.Sh HISTORY
.Nm
first appeared in
.Ox 1.2 .
.Pp
A
.Nm makekey
command appeared in
.At v7 .
.Pp
The
.Fl S
option first appeared in
.Mx 11 .


src/usr.bin/encrypt/encrypt.c
/*	$OpenBSD: encrypt.c,v 1.21 2004/07/13 21:09:48 millert Exp $	*/

/*
 * Copyright (c) 2008
 *	Thorsten Glaser <tg@mirbsd.de>
 * Copyright (c) 1996, Jason Downs.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR(S) ``AS IS'' AND ANY EXPRESS
 * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR(S) BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/types.h>
#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <login_cap.h>

__RCSID("$MirOS$");

/*
 * Very simple little program, for encrypting passwords from the command
 * line.  Useful for scripts and such.
 */

#define DO_MAKEKEY 0
#define DO_DES     1
#define DO_MD5     2
#define DO_BLF     3
#define DO_CLASS   4
#define DO_SAME    5

extern const char *__progname;
char buffer[_PASSWORD_LEN];

void	usage(void) __dead;
char	*trim(char *);
void	print_passwd(char *, int, void *);

void
usage(void)
{

	(void)fprintf(stderr,
	    "usage: %s [-k] [-b rounds] [-c class] [-m] [-[Ss] salt] [-p | string]\n",
	    __progname);
	exit(1);
}

char *
trim(char *line)
{
	char *ptr;

	for (ptr = &line[strlen(line)-1]; ptr > line; ptr--) {
		if (!isspace(*ptr))
			break;
	}
	ptr[1] = '\0';

	for (ptr = line; *ptr && isspace(*ptr); ptr++)
		;

	return(ptr);
}

void
print_passwd(char *string, int operation, void *extra)
{
	char msalt[3], *salt = NULL;
	login_cap_t *lc;
	int pwd_gensalt(char *, int, login_cap_t *, char);
	void to64(char *, int32_t, int n);

	switch(operation) {
	case DO_MAKEKEY:
		/*
		 * makekey mode: parse string into separate DES key and salt.
		 */
		if (strlen(string) != 10) {
			/* To be compatible... */
			errx(1, "%s", strerror(EFTYPE));
		}
		strlcpy(msalt, &string[8], sizeof msalt);
		salt = msalt;
		break;

	case DO_MD5:
		strlcpy(buffer, "$1$", sizeof buffer);
		to64(&buffer[3], arc4random(), 4);
		to64(&buffer[7], arc4random(), 4);
		strlcpy(buffer + 11, "$", sizeof buffer - 11);
		salt = buffer;
		break;

	case DO_BLF:
		strlcpy(buffer, bcrypt_gensalt(*(int *)extra), _PASSWORD_LEN);
		salt = buffer;
		break;

	case DO_DES:
		salt = extra;
		break;

	case DO_CLASS:
		if ((lc = login_getclass(extra)) == NULL)
			errx(1, "unable to get login class `%s'",
			    extra ? (char *)extra : "default");
		if (!pwd_gensalt(buffer, _PASSWORD_LEN, lc, 'l'))
			errx(1, "can't generate salt");
		salt = buffer;
		break;

	case DO_SAME:
		salt = extra;
		break;
	}

	(void)fputs(crypt(string, salt), stdout);
}

int
main(int argc, char **argv)
{
	int opt;
	int operation = -1;
	int prompt = 0;
	int rounds;
	void *extra = NULL;		/* Store salt or number of rounds */

	if (strcmp(__progname, "makekey") == 0)
		operation = DO_MAKEKEY;

	while ((opt = getopt(argc, argv, "b:c:kmpS:s:")) != -1) {
		switch (opt) {
		case 'k':                       /* Stdin/Stdout Unix crypt */
			if (operation != -1 || prompt)
				usage();
			operation = DO_MAKEKEY;
			break;

		case 'm':                       /* MD5 password hash */
			if (operation != -1)
				usage();
			operation = DO_MD5;
			break;

		case 'p':
			if (operation == DO_MAKEKEY)
				usage();
			prompt = 1;
			break;

		case 'S':			/* BSD crypt (all algos) */
			if (operation != -1)
				usage();
			operation = DO_SAME;
			extra = optarg;
			break;

		case 's':                       /* Unix crypt (DES) */
			if (operation != -1 || optarg[0] == '$')
				usage();
			operation = DO_DES;
			extra = optarg;
			break;

		case 'b':                       /* Blowfish password hash */
			if (operation != -1)
				usage();
			operation = DO_BLF;
			rounds = atoi(optarg);
			extra = &rounds;
			break;

		case 'c':                       /* user login class */
			if (operation == DO_SAME)
				usage();
			extra = optarg;
			operation = DO_CLASS;
			break;

		default:
			usage();
		}
	}

	if (operation == -1)
		operation = DO_CLASS;

	if (((argc - optind) < 1) || operation == DO_MAKEKEY) {
		char line[BUFSIZ], *string;

		if (prompt) {
			if ((string = getpass("Enter string: ")) == NULL)
				err(1, "getpass");
			print_passwd(string, operation, extra);
			(void)fputc('\n', stdout);
		} else {
			/* Encrypt stdin to stdout. */
			while (!feof(stdin) &&
			    (fgets(line, sizeof(line), stdin) != NULL)) {
				/* Kill the whitesapce. */
				string = trim(line);
				if (*string == '\0')
					continue;
				
				print_passwd(string, operation, extra);

				if (operation == DO_MAKEKEY) {
					fflush(stdout);
					break;
				}
				(void)fputc('\n', stdout);
			}
		}
	} else {
		char *string;

		/* can't combine -p with a supplied string */
		if (prompt)
			usage();

		/* Perhaps it isn't worth worrying about, but... */
		if ((string = strdup(argv[optind])) == NULL)
			err(1, NULL);
		/* Wipe the argument. */
		memset(argv[optind], 0, strlen(argv[optind]));

		print_passwd(string, operation, extra);

		(void)fputc('\n', stdout);

		/* Wipe our copy, before we free it. */
		memset(string, 0, strlen(string));
		free(string);
	}
	exit(0);
}
